### DISCUSSION TITLE
KlikGen  firmware for akai?  Mpc Magic?

### DISCUSSION THREAD

**Discussion by ModulationNation** [TheKikGen/MPC-LiveXplore#56] [Ideas]
Hi, it's me again!   I talked my self into a solution while over on a KlikGen discussion.  So I moved here...  Can you port KlikGen to the MPC Live?  if i read correctly from the KlikGen repo you've written a programable transport layer protocol for midi inside of sysex? heavy!!!  i think using two akai devices running Force software controlling midi magic.would be very interesting..   was the # of independent tempos based on the # processors?  could it be replicated or exceeded on Akai? did i read buffered?  could we get more Clocks?  i'd even be happy to run one headless.  just sysex a shell and the midi interface would work for me! but whatevers possible lets try it!

**Comment by ModulationNation** [comment]
HEY FAM
>
ok so I have yet to get ssh access into theKikGen on my mpc live.  I've tried off and on for a couple months.   it ask for a password? the only wiki entry i could find said there is none.  leaving it blank doesn't work.  i've read most of the files in the bootstrap(though i only understand comments).   I've tried adding a rsa.pub to the Authorized_host file.  I've tried to edit the ssh.config.   I've gotten iamforce2@NONE to launch(through reading files).  still no ssh.
>
I have a force running mockbaMod, I've set up the configs for that, and i have that working flawlessly.  I think I'm most interested in the armbian dev img for the mpc.  but my trouble thus far getting ssh access has me cautious.  my end goal is to figure out the sysex trickery I know is possible from the files on your various repos.  
>
>MY SETUP<
>Force(mockbaMod)
>MPC(iamforce)
>iconnectivity MioXL
>roland Alpha juno2
>yahmaha DX7
>Behringer xair(mixer)
>behringer xtouch(control surface)
>roland FC-300(midi pedal)
>Behringer ms1/modelD/crave/pro1
>
 I'd like to have the mpc and force set up together.  I'D like access to the public and private midi cards on both devices.  mostly via the mpcLive.  I'd like the force(MM) to run pretty much how it is, sequencing  hardware, i like how MM lets me automatically generate songs by copying/pasting resizing/respeeding clips via note/CC functions.  on the mpc I would like to mostly be doing sysex.  sequencing sysex. launching/recording sysex streams.   is there a way to setup time based .syx files? or a Note/CC to sysex translator  config file?  Batch translation  is what I want.   datapipe bit flipping?   gating with the intentional use/misuse of the midi sysex id bug.   rerouting, and remapping controlled by LFO's.     mpc sending data as other devices on my network.   like switching routing and filtering on the mioxl(it has preset selection, so using sysex to make changes on the fly is probably overkill, but hey why not).    I'm trying to also build/install Ctrlr on either my macbook/linux desktop.   didn't have much luck on macos.  starting to try on linux, but i'm certain i should be setting up a dev environment in a specific way that i'm unaware of(file structures, symlinks).   I'm assuming codeblocks and JUCE are involved.   end goal is for the light pads on the force to interacting with euclidian/algorithmic sequencers on the mpc  which in return triggered clips and mockba functions to the force.   and LFO's/clips controlling MioXL port filters/remaps.   LFO/clip driven sysex gates.  CC>sysex for Juno/DX7
>
>
Are the problems i'm having with ssh, user ignorance?  or is the Documentation behind development?
>
>
Am i in over my head, is it possible to implement this type of setup with zero programming background?
>
>
is this possible for me to achieve.
>
>
I think If I could figure out how to implement this myself I could teach others. I'd love to be involved with documentation for these projects.  but right now i feel like i' beating my head against the wall.  not sure if i ca even accomplish it , let alone teach it>
>
>thanks for your time and work, any help towards my end goal would be much appreciate<

**Reply by ModulationNation** [reply]
i'm moving forwards towards understanding your code...  i understood from the comments throughout the code, exactly  what you were doing.  still wrapping my head around the how.  i found an indexable github wiki page.  one page you were explaining and directly referencing the most interesting bits of both  the MPCLive, and kikpad code, in reference to the public and private midi ports...  i got overly excited when i first started reading through the 4x4kik repos...  the sysex file!!!!  i had been looking at things like pd, and supercolider for 56 hrs before i thought to to put github in the search feild..  the a realtime programable midi transport protocol written into unregistered sysex... blew my mind... as well as the idea of intentionally losing/dropping data...  to gate signal at midi out, feels a lot more like modular, than just gating midi...   
>
>
still can't get into ssh...   I'm baffled on that one....  i've had thekikgen image on my machine for months.  i've gotten iam force to boot reliably but no ssh..

**Comment by TheKikGen** [comment]
Hi. it is totally possible to implement a "new" sysex protocol for the Force:MPC, and I think it is a really good idea, as the native sysex is not documented at all.  The approach is more or less the same as what I did with IamForce or other libs : man in the middle :
Midi event -> [my module] -> MPC.
In the same principle, I added some CC to IamForce to generate specific behaviours like PLAY, REC, STOP, NEXT SEQUENCE, etc...
https://github.com/TheKikGen/MPCLiveXplore-libs/blob/df1be83ec209a1e8be63759bb05fa217b4f0d0ae/src/tmm-plugins/tmm-IamForce2.c#L80
and here :
https://github.com/TheKikGen/MPCLiveXplore-libs/blob/df1be83ec209a1e8be63759bb05fa217b4f0d0ae/src/tmm-plugins/tmm-IamForce2.c#L882
